bootstrap/kubeadm/internal/controllers/kubeadmconfig_controller.go

```diff
@@ -326,6 +325,20 @@ func (r *KubeadmConfigReconciler) refreshBootstrapTokenIfNeeded(ctx context.Cont
 
 	secret, err := getToken(ctx, remoteClient, token)
 	if err != nil {
+		if apierrors.IsNotFound(err) {
```

Comment:
I think we should limit this behaviour (re-create the token) to only when configOwner.IsMachinePool() or when the config owner is a machine and it doesn't have the data secret field set.

This is required because machines, once data secret field is set, are never going to pickup up new data secrets (and having re-created by un-used data secrets around will be noisy/possibly confusing when triaging issues).

We also using test coverage for this change

Reply:
@fabriziopandini  Any advice for the issue I met? Current logic bug here is:
1. Token is created and tracked by cluster-api controller, but the lifecycle is handled by workload cluster, it will be deleted by workload cluster without info cluster-api controller
2. Machine Pool is created by cluster-api controller, but cluster-autoscaler is used to trigger the replica number expected for the pool

So whenever the cluster is `paused`, things may be out of control, and we can see node join issue, especially when we use the default `15 mins` for token ttl.

In our case, I have changed the token ttl to 7 days to avoid the issue, but still I think we should find a way to handle it. If auto refresh is not recommended, will it be acceptable that we got some kind of ctl to do the refresh?

Reply:
What is the reason for pausing the cluster? Maybe when pausing the cluster, you also should pause autoscaler for this cluster in some way?!

Reply:
In our case: 
We have two controllers cluster, so we can migrate cluster to another controller cluster during the migration, which will pause the cluster. In meet any issue during migration, the time for pause may be longer.

As I said, we have use longer til to solve this. But still think this issue should be handled.

Cluster api support paused feature, then I think it's reasonable to handle this kind of case. Just my opinion😄

Reply:
While the approach you mention here is valid, the feedback from Fabrizio 

> limit this behaviour (re-create the token) to only when configOwner.IsMachinePool() or when the config owner is a machine and it doesn't have the data secret field set.

should still apply — given that we don't want to renew every token regardless if the owner has joined or not

Reply:
The issue I met is for machinepool, so if I limited the refresh  token when not found to machinepool created by cluster-api, can it be acceptable?

Reply:
Sure, it is ok to improve how MP recovers after a Cluster is paused for a long time.
What we want to make sure is that the current change doesn't impact anything which isn't owned by a MP.

Also, please do not consider pause as a regular Cluster API feature. 
It is an option that we introduced for allowing extraordinary (emphasis on extraordinary) maintenance operations and it is assumed deep knowledge of the system for whatever happens when the cluster is paused.

Reply:
I've covered this initial request (only recreate for machine pools) in my dupe PR https://github.com/kubernetes-sigs/cluster-api/pull/11520. Sorry that I didn't find this PR earlier – maybe I got distracted by closed PRs or so. Our PRs are now quite similar.

Reply:
@AndiDog  Great to know,  and you have got all the test ready, please go on with your PR, nice to have the issue be fixed soon